Manual: New chapter on polymorphic types and functions in the tutorial

Octachron opened the pull request:
This PR proposes to add a new chapter to the manual to describe in more detail the properties of polymorphic functions and type in OCaml. More precisely, the new chapter would describe:
- the relaxed value restriction: starting from weakly polymorphic types to the relaxed value restriction itself
- polymorphic recursion and the use of explicit polymorphic annotation
- higher-rank polymorphic functions using explicitly polymorphic record fields or object methods.

The main objective of this PR is to gather in the manual the information that is already available on the web but not always easily discoverable.

As an important side-effect, this PR moves the description of explicitly polymorphic type annotation out of the extension section of the language reference manual.

Note: this is still a work in progress, comments and criticisms are more than welcome.

Octachron committed:
Manual: new section on polymorphism limitations

Octachron committed:
Manual: move polymorphic annotation out of exten

objmagic commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+undecidable; therefore using this kind of higher-rank function requires to
+handle manually these universally quantified type.
+
+In OCaml, there is three main way to introduce this kind of explicit universally
```
there are three main ways

objmagic commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+
+This distinction between weakly and generic polymorphic type variable protects
+OCaml program from unsoundness and runtime errors. To understand from where 
+unsoundness might come, consider this simple function which swap a value "x"
```
swaps

objmagic commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+\begin{caml_example}
+let id_again = fun x -> (fun x -> x) (fun x -> x) x;;
+\end{caml_example}
+Whith this argument, "id_again" is seen as a function definition by the type
```
with

objmagic commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+Allied with this relaxed value restriction, type parameter covariance
+helps to avoid eta-expansion in many situation.
+
+\subsection{Asbract data types}
```
abstract

Octachron replied:
Thanks for noticing these errors.

Octachron committed:
Fix: spelling/grammatical/whitespace errors

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+This chapter details each of these situations and, if it is possible,
+how to recover genericity.
+
+\section{Weak polymorphism and mutations}
```
mutation

Octachron replied:
Fixed.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+
+More generally, weak types protect the program from undue mutation of
+values with a polymorphic type. Moreover, weak type cannot appear as the type
+of toplevel values: the type of values must be known at compilation time.
```
You _could_ point out that this would be problematic if two distinct modules relied on this module's interface, unifying exported weak types to different things. You _should_ spell out the error message in entirety (for example, by using an example that raises it), because it's confusing when you first see it and it would be a good anchor for people looking for the error message text.

Octachron replied:
Good points. I have also added a remark on the fact that this error cannot be repoduced in the toplevel interpreter to avoid potential confusion.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+There is another partial solution to the problem of unnecessary weak type
+in the type of value, which is implemented directly within the type
+checker. Briefly, it is possible to prove that weak types that only appear
+as type parameters in a covariant position can be safely generalized to
```
People sometimes use "positive position" instead of "covariant position", and it makes as much sense when subtyping is not considered. It would be helpful to refer to this other wording at some point.

Octachron replied:
I have added a short mention. It may be better to weave the name more cohesively in the text.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+\begin{caml_example}
+  type 'a tree = Leaf of 'a | Branch of 'a tree * 'a tree;;
+\end{caml_example}
+is covariant in its parameter "'a" if it preserves the order of the
```
I feel that this description is too abstract. (preserves the order of the subtyping relationship?).

Octachron replied:
You are right, I fear that I let out too much of my inner mathematician here. Anyway, the whole section have been rewritten (see below).

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+Then, if the type constructor "'a t" is covariant in "'a",
+the subtyping relationship "x :> xy" can be lifted to "x t :> xy t".
+
+This notion of covariance can be used to prove that a weak type"'_a" that
```
Another explanation would be that as soon as a type variable appears in a position describing mutable state, such as `'a ref`, it becomes invariant, and that thus covariant variables will never denote mutable locations and can be safely generalized.

I don't know whether we need to have both explanations. I feel that you are repeating a lot of what was rather clearly written in Jacques Garrigue's article, and that it may be enough to more concisely explain the behaviour and intuition, and refer to the article for people looking for an in-depth explanation.

(Also that could help in keeping the documentation less abstract/formal.)

Octachron replied:
I have tentatively rewritten the variance section to put more emphasis on concrete examples. Linking directly invariance with mutable state seems to give a better flow to the text.
I would also add a link to Jacques Guarrigue's article later.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+
+module Implementation = struct
+  type 'a t = { get: int -> 'a; len:int }
+  let empty ()= let a = [||] in { get = (fun n -> a.(n) ); len = 0 }
```
That code is too complex; what about just using lists under the hood? (Or `int -> 'a option` if you want something exotic)

Octachron replied:
Sure.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,387 @@
+helps to avoid eta-expansion in many situation.
+
+\subsection{Abstract data types}
+Moreover, when the associated type definitions are visible, the type checker
```
Let's avoid "associated type" which means something else. You could just say "When the type definitions are exposed".

Octachron replied:
Fixed.

gasche commented:
I think that's a good idea. I reviewed the part before "Polymorphic recursion" so far. I think it is important to keep things simple, concrete and example-based as soon as we can, and avoid formalism -- because it can turn people away and make the document feel less accessible.

Octachron committed:
Fix spelling error

Octachron committed:
Fix: non-generalizable type error example

Octachron committed:
Fix: variance subsection

Octachron committed:
Fix: variance and abstract types

Octachron committed:
Fix grammar and add todo for indentation

Octachron commented:
@gasche 
Thanks for your time and review.

bluddy commented:
Thank you so much for doing this!

Octachron committed:
Fix: miscorrection in section 5.1.5

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,435 @@
+Error: The type of this expression, '_a option ref,
+       contains type variables that cannot be generalized
+\end{verbatim}
+In this case, the error can be solved by either fixing the relevant
```
`fixing` suggests that something was "broken" before, you should rather use something like

> To solve this error, it suffices to add an explicit type annotation to make the type precise at declaration time.

I would also add:

> This is in any case a good practice for such global mutable variables; otherwise, it will "adopt" the first type of use, which can give confusing type errors if there is a mistake at the first use site -- later, correct uses would be flagged as errors.

Octachron replied:
Good points. I have adopted your new paragraph with some stylistic modification.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,435 @@
+let option_ref: int option ref = ref None
+\end{verbatim}
+or by transforming "option_ref" in a function%
+\begin{verbatim}
```
I'm not sure it makes sense to show the two solutions on the same example. It seems pretty clear when I read `let foo = ref None` that a mutable global variable is intended to be initialized later, and making it a function changes its semantics in a way that seems rather artificial to me. Later examples will propose eta-expansion as a solution to get more polymorphism, so maybe you could just this ", or...." part of the paragraph.

Octachron replied:
Yes, skipping the second point is probably a good idea since it does not really make sense with this specific example.

gasche commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,435 @@
+\begin{verbatim}
+let option_ref () = ref None
+\end{verbatim}
+An important point is that, by nature, this error cannot appear in OCaml
```
I think we could skip that paragraph as well -- I'm surprised I notice how to shorten other people's texts when mine are always so verbose. The explanation you give is not very good¹, and in any case people are rarely surprised when stuff works, they want explanations when it does not.

¹: The reason why weak polymorphic values are forbidden at module interfaces is that the module may be linked twice with different unifications, making it unsound. On the other hand, a toplevel declaration cannot be used by two consumers simultaneously, there is a single "continuation" which is the rest of the toplevel section. The sentence below ("in the toplevel, ...") is not really an explanation.

Octachron replied:
You are right the explanation is bad.

Anyway, I was worried about people trying to reproduce the error message in a toplevel after reading the previous paragraph. But it is probably a too unfrequent occurence to warrant any explanation.

Octachron committed:
Manual: improve non-generalizable error paragraph

damiendoligez commented:
@objmagic @gasche Are you done reviewing?

gasche commented:
I'm not done, unfortunately. I need to invest more time and there are in places some reformulation that I'm thinking of proposing (instead of just commenting). Are you asking because you would like to merge soon? What's the calendar for when this should be settled?

(This manual change is more ambitious than others and it also hits closer to home in terms of discussing non-trivial points of advanced type system / language design. I could see myself falling into an unbounded process of gradual refinement (especially if we get Didier involved, which would be a good idea). I think that we should instead aim for a mergeable version that is reasonable, and refine later if we want to -- le mieux est l'ennemi du bien -- but I haven't read the whole thing yet so I cannot say I'm there.)

@Octachron what's your own feeling/preference? Does it feel "ready" to you, do you have specific expectations in term of perfectionism level vs. merge time?

Octachron commented:
For this specific PR, my objective is more to reach a good first step that could be readily improved upon in the future, either directly or by collecting feedback. I readily admit that with the limited depth of my understanding on some of the subjects covered trying to get it right on the first version would be overly ambitious.

However, there are still rough parts, spelling errors and some unnecessary complexity
in the current version, that I would like to correct (by the end of the week if needed?). I also would not mind some external confirmation that there is no factual errors in the later part of the text.

@gasche, if you find the time to write these propositions, I would be interested to merge them.

p.s. : 

@damiendoligez , @gasche: As an echo to the trunk manual discussion, this is another situation where I feel that it might be nice to have an experimental web version of the manual to gather feedback more easily.

Octachron committed:
GPR#594 fix: grammar

Octachron committed:
GPR#594: link to Jacques Guarrigue's article

Octachron committed:
GPR#594: improved polymorphic recursion example

Octachron committed:
GPR#594: improved higher-rank polymorphism

damiendoligez commented:
@gasche There's no calendar, I was just pinging the PR to see if it was ready.

Octachron committed:
Minor fixes

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+polymorphic type variables, sometimes shortened as weak type variables. A weak
+type variable is a placeholder for a single type that is currently unknown.
+Once the specific type "t" behind the placeholder type "'_a" known, all
+occurrences of "'_a" will be replaced by "t".
```
Give an example here, e.g.

```
let store = ref None
  store : `_a option ref
store := Some 1
  store : int option ref
```

Octachron replied:
Thanks for the suggestion, giving an example here improve nicely the flow of the section.

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+undue mutation of values with a polymorphic type.
+
+%todo: fix indentation in pdfmanual
+Moreover, weak type cannot appear as the type of toplevel values:
```
values in toplevel modules

Octachron replied:
I would rather say that weak types cannot appear in signatures in order
to include cases like

```
module M = struct module N = struct let store = ref None end end
```

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+replaced by weak types in a modular way is a difficult question. A first
+natural idea would be to state that any mutable object like reference, array
+or a record mutable field should never have a polymorphic type. Unfortunately,
+this rule is not sufficient, as illustrated by the following function that
```
Stumbling as reader over this. Better: "this rule is not only impractical but also not even a sufficient solution to the problem, as illustrated ..."

Octachron replied:
Thanks for your remark, I did not notice that this hand-wavy description fits multiple
type system (some which are even sound as far as I can tell). So it is probably better
to delete my babbling and shows the problem directly. For instance, 

> Identifying the exact context in which polymorphic types should be
> replaced by weak types in a modular way is a difficult question. Indeed
> the type system system must take in account the fact that functions
> can hide persistent mutable state, as illustrated ...
> At the same time, it should be possible to use a local mutable state without
> impacting the type of a function.

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+The value restriction combined with this generalization for covariant type
+parameters is called the relaxed value restriction.
+
+%question: is here the best place for describing variance?
```
Avoid variance in the main text, e.g. "In simple words, the generalization can also be done when the type only refers to values returned by functions that never appear as arguments of other functions (i.e. the values are all in covariant positions)"

Octachron replied:
I am not sure that it is better to avoid the word variance here. I fear that any reasonable approximation would end up quite wordy:

> "As a first approximation, the generalization can also be done when the type only refers to values returned by functions that never appear as arguments of other functions or parameter of a mutable type constructor."

Is it not simpler to use the name "variance" and then explains what is variance? Note that it may be well my academic background speaking here.

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+
+\section{Polymorphic recursion}\label{s:polymorphic-recursion}
+
+The second major class of non-genericty is directly related to the problem
```
typo: genericity

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+The second major class of non-genericty is directly related to the problem
+of type inference for polymorphic function. In some circumstances, the type
+inferred by OCaml might be not general enough to allow the definition of
+some recursive functions on recursive type.
```
Before going on with this "brittle" stuff, show the reader safe grounds. Explain what a regular ADT is before going on with non-regular types. It is important to see the difference.

Octachron replied:
Well, up to now, I was not fully aware of the distinction between regular and non-regular ADT. Thanks for the pointer! And you are right making the distinction will make the origin of the problem much clearer for the reader.

gerdstolpmann commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,443 @@
+non-polymorphic type:
+\begin{caml_example}
+  let sum: 'a 'b 'c. 'a -> 'b -> 'c = fun x y -> x + y;;
+\end{caml_example}
```
I am missing here two "footnotes". First, the practical comment that many types in explicitly annotated functions can be simply left as "_", e.g. `let rec depth : 'a . 'a nested -> _`, limiting the annotation effort.

Second, point out that there is also a variation using "newtypes" (`let rec depth : type t . t nested -> _`) which often yields the same effect though the treatment of the variable inside the function is slightly different when GADTs are involved (i.e. the variable can be specialized differently depending on the GADT case whereas the normal type variable `'a` means always the same type on the current recursion level).

Octachron replied:
> First, the practical comment that many types in explicitly annotated functions can be simply left as "_", e.g. let rec depth : 'a . 'a nested -> _, limiting the annotation effort.

Good point.

> Second, point out that there is also a variation using "newtypes" (let rec depth : type t . t nested -> _) which often yields the same effect though the treatment of the variable inside the function is slightly different when GADTs are involved (i.e. the variable can be specialized differently depending on the GADT case whereas the normal type variable 'a means always the same type on the current recursion level).

It is certainly an important point to make and this is the right place to do so, but I fear this not the right time since GADT and locally abstract type are still relegated to the extension section of the manual.

Octachron committed:
GPR#594: simple example for weak type unification

Octachron committed:
GPR#594: toplevel signature and weak types

Octachron committed:
GPR#594: introduction for value restriction

Octachron committed:
GPR#594 typo and grammar for weak type section

Octachron committed:
GPR#594: minor, relaxed v.r. section

Octachron committed:
GPR#594: regular and non-regular types

Octachron committed:
GPR#594: optional type annotations

Octachron committed:
GPR#594: minor fixes, polymorphic recursion

Octachron committed:
GPR#594: minor fixes, higher-rank polymorphism

Octachron commented:
I have updated the chapter to take in account @gerdstolpmann comments; thanks again.
As a side note, I have also an [experimental html page](http://www.polychoron.fr/ocaml-nonmanual/polymorphism.html) for the chapter,  for potentially interested readers that were detered by the raw latex (and extensions) files.

nojb commented:
The page looks great, except that the table of contents overlaps with the main content.  Using Safari on OS X 10.11.

Octachron commented:
I have just fixed this specific rendering issue on chrome. I am not completely sure if it works now with Safari.

nojb commented:
Yes, it works.  Thanks!

Octachron committed:
Manual: expect annotation for polymorphism.etex

Octachron committed:
Minor wording update

Octachron committed:
update caml_example to use toplevel mode

Octachron committed:
Update to use the new weak type prefix

alainfrisch commented:
I think we should try to merge this nice contribution before it's forgotten.

@gasche What's your opinion?  Do you think we can merge in the current form?  The text can always be improved later on.

damiendoligez commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+weak type variables.
+A weak type variable is a placeholder for a single type that is currently
+unknown. Once the specific type "t" behind the placeholder type "'_weak1"
+known, all occurrences of "'_weak1" will be replaced by "t". For instance,
```
Missing word: "**is** known"

Octachron replied:
Fixed.

damiendoligez commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+let two = swap store 3;;
+\end{caml_example}
+After these three swaps the stored value is "3". Everything is fine up to
+now. We can then try to swap "2" with a more interesting value, for
```
Why "2"? Should be "3" no?

Octachron replied:
Indeed.

damiendoligez commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+\subsection{Variance and value restriction}
+Variance describes how type constructors behave with respect to subtyping.
+Consider for instance a pair of type "x" and "xy" with "xy" a subtype of "x",
+i.e "xy :> x":
```
Don't you mean "supertype" here?

Octachron replied:
There was indeed an order inversion problem here. I choose to replace it by `"x" a subtype of "xy"` to avoid introducing the notion of `supertype`.

garrigue replied:
I suppose he's using OCaml cast syntax here: `(_ : xy :> x)`. Might require some explanation.

Octachron replied:
I am indeed using OCaml cast syntax. Thinking about it, it may be better to only use words, i.e. `is a subtype` rather than symbolic notation.

damiendoligez commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+mutable state it becomes invariant. As a corollary, covariant variables will
+never denote mutable locations and can be safely generalized.
+For a better description, interested reader can consult the original
+article by Jacque Guarrigue on
```
Jacques Garrigue

Octachron replied:
Fixed.

gasche commented:
@alainfrisch unfortunately I'm not sure I will have any time to review the rest of this PR before the release. I would be happy to trust you (and/or @damiendoligez, etc.) and merge if you are happy with the current state.

alainfrisch commented:
I've looked only superficially, so I'll @damiendoligez decides.

damiendoligez reviewed (changes requested):
It'll be good to merge when my few comments (above) are addressed.

Octachron committed:
Various fixes after Damien review

Octachron committed:
Fix few academic idiosyncrasies

Octachron committed:
More grammatical error fixes

Octachron committed:
GPR#594: update changes

Octachron committed:
add a type annotation to simplify an example type

Octachron commented:
I have fixed the issues raised by @damiendoligez , corrected few grammatical errors or idiosyncrasies along the way and added a Changes entry.

garrigue commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+For a better description, interested reader can consult the original
+article by Jacque Guarrigue on
+\url{http://www.math.nagoya-u.ac.jp/~garrigue/papers/morepoly-long.pdf}
+%question: what is the best url for this article?
```
That's the canonical one.

Octachron replied:
Thanks for confirming.

garrigue commented on manual/manual/tutorials/polymorphism.etex:
```diff
@@ -0,0 +1,478 @@
+``for all type "'a", "depth" maps "'a nested" values to integers''.
+Whereas the standard type "'a nested -> int" can be interpreted
+as ``let be a type variable "'a", then "depth" maps "'a nested" values
+to integers''. There is two major differences with these two type
```
`There are two major differences between`

Octachron replied:
Fixed, thanks.

damiendoligez commented:
Good to go, and we can always enhance it later.

Should we squash the commits before merging?

Octachron committed:
Fix: Jacques Garrigue review

Octachron committed:
remove forward reference to explicit polymorphism

since explicit polymorphism annotations have been moved to the standard
language reference chapter.

Octachron commented:
I think that squashing the commits is probably better for this PR, since the majority of commits after the first one are small fixes with an useless commit message for an external reader.

gasche reviewed (approved):
Approved (on the basis of Damien's more exhaustive review). @Octachron, would you do the work of squashing, merging and cherry-picking in 4.06?

Octachron commented:
@gasche : Certainly, I will.

Octachron commented:
Merged and cherry-picked on 4.06(641dab319e7b3f75a).

gasche commented:
Thanks! And thank for your work on the manual in general.

Something I've been thinking about but haven't got around to doing it yet is to make a pre-release of the 4.06 manual in a temporary place, to include in 4.06-preparation-related announcements. If you also think this could be of use, would you be interested in taking care of it? (I have web space where it's easy to push temporary stuff; if you don't, I will handle the whole thing).

Octachron commented:
I think releasing a beta version of the manual is a good idea. If it just a matter of putting a 4.06 version of the manual online, I could do the same thing as http://www.polychoron.fr/ocaml-nonmanual/mpr7497/ and make it available at http://www.polychoron.fr/ocaml-beta-manual/4.06/ with a slightly adjusted disclaimer text (and keeping the "alternative" color scheme).

gasche commented:
Yes, I think that would be very nice. You could advertise it in the email thread of the latest beta announcement, and possibly point to specific part of the manuals that changed (typically this new section).

Drup commented:
@gasche Couldn't we just have a version of the manual for trunk automatically uploaded ? (http://caml.inria.fr/pub/docs/manual-ocaml-trunk/ , for example)

gasche commented:
We certainly could, but someone has to do the work.
